src/devices/SoftPwm/SoftwarePwmChannel.cs

```diff
@@ -105,106 +110,89 @@ public SoftwarePwmChannel(int pinNumber, int frequency = 400, double dutyCycle =
         }
 
-        private void RunSoftPWM()
+        private void Run()
         {
-            if (_usePrecisionTimer)
+            while (!_isTerminating)
             {
-                Thread.CurrentThread.Priority = ThreadPriority.Highest;
-            }
-
-            while (_runThread)
-            {
-                // Write the pin high for the appropriate length of time
-                if (_isRunning)
+                if (!_isRunning)
                 {
-                    if (_pulseWidthInMilliseconds != 0)
-                    {
-                        _controller.Write(_pin, PinValue.High);
-                        _isStopped = false;
-                    }
-
-                    // Use the wait helper method to wait for the length of the pulse
-                    if (_usePrecisionTimer)
-                    {
-                        Wait(_pulseWidthInMilliseconds);
-                    }
-                    else
-                    {
-                        Task.Delay(TimeSpan.FromMilliseconds(_pulseWidthInMilliseconds)).Wait();
-                    }
-
-                    // The pulse if over and so set the pin to low and then wait until it's time for the next pulse
-                    _controller.Write(_pin, PinValue.Low);
+                    Thread.Yield();
+                    continue;
+                }
 
-                    if (_usePrecisionTimer)
-                    {
-                        Wait(_periodInMilliseconds - _pulseWidthInMilliseconds);
-                    }
-                    else
-                    {
-                        Task.Delay(TimeSpan.FromMilliseconds(_periodInMilliseconds - _pulseWidthInMilliseconds)).Wait();
-                    }
+                if (_pinHighTime != TimeSpan.Zero)
+                {
+                    _controller.Write(_pin, PinValue.High);
+                    Wait(_pinHighTime);
                 }
-                else
+
+                if (_pinLowTime != TimeSpan.Zero)
                 {
-                    if (!_isStopped)
-                    {
-                        _controller.Write(_pin, PinValue.Low);
-                        _isStopped = true;
-                    }
+                    _controller.Write(_pin, PinValue.Low);
+                    Wait(_pinLowTime);
                 }
             }
+
+            _controller.Write(_pin, PinValue.Low);
         }
 
         /// <summary>
         /// A synchronous wait is used to avoid yielding the thread
         /// This method calculates the number of CPU ticks will elapse in the specified time and spins
         /// in a loop until that threshold is hit. This allows for very precise timing.
         /// </summary>
-        /// <param name="milliseconds">The milliseconds to wait for</param>
-        private void Wait(double milliseconds)
+        /// <param name="ticks">The time to wait.</param>
+        private void Wait(TimeSpan timeout)
```

Comment:
The wait logic was just centralized here. `_stopwatch` being initialized determines if we are high precision or not and we `Sleep` if we are not high precision. Using `Task.Delay` isn't really worthwhile if we aren't actually async.